# Supervivencia Titanic


Orígen de datos: https://www.kaggle.com/c/titanic


<img src="./titanic.jpg">

## Librerías a utilizar

In [1]:
import numpy as np
import pandas as pd

from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.svm import SVC
from sklearn.neighbors import KNeighborsClassifier

## Importar datos
- Descargar datos a utilizar desde la web
- Guardar los datos para que queden disponibles
- Importar los ficheros .csv con los datos

In [3]:
# Importar datos a utilizar desde la web
# url_train = "https://storage.googleapis.com/kaggle-competitions-data/kaggle/3136/train.csv?GoogleAccessId=competitions-data@kaggle-161607.iam.gserviceaccount.com&Expires=1504325799&Signature=ZREfojvjBxGz2%2FF4Yu1Kka3j%2BylZLrFiNNgzVOjRJk4XOpKFgV%2BK4eM0fS874l3fXVvg051nUB%2BDsCVORI%2FN3Cgt%2FyE5OLE5zpwpe%2Ff1zWD5JPS91S767g0oRCT%2FEwqfV4Dx2TwIT1dEZXVBfLVL5qlp5lIDJSOnfezs3at%2BhMpGDJUIKzbtnbA%2BLn%2FhRbtZqa%2B8SnhdmwnDOSkxA9ZYDK9QpmITb7liBRgd3DthiTNCp2ajG%2FSwnOgV8Tm%2FNefDKTMtbRgljgMzvpuQ031SomV5iBu8WmIygqxjfib6DhKP4m5t%2FahZzBYIEfY7AmBokiZDNlwtVwOivWDANIXBzA%3D%3D"
# url_test = "https://storage.googleapis.com/kaggle-competitions-data/kaggle/3136/test.csv?GoogleAccessId=competitions-data@kaggle-161607.iam.gserviceaccount.com&Expires=1504325924&Signature=hL1wMTmPKV5OYjZqHckFPzc%2BWAbEp0dD4AtzxgaLQXosZaj4dtGlM2PEl0Jq0WNTAZhUIQifGweKJM6qTa8G%2FQgRyRunN%2BQa2hqocGDsQ77wi71CkmOvqKogZL0fMLnU%2Bowd7ww5OBSerpAYXY%2FC8vT5J%2BrVfMAxfXQfyO9w1as%2BV7zUia14TfplOxMJwJQGNib08T8qIpZYjyvSZZcWEvoIKPqpnZBqDqnrxGOOvm%2BDpvWL04ftIg7mIAFOlhxiJ87X%2BQOAHyv9u9HBia2lbjn6h%2BoI8HmF0o6i%2BZs6BK3EyMMEEaE1G5dv02j%2B8%2B%2F9V%2FWukLam2y%2Fx1jMs6D3wKQ%3D%3D"
# df_train = pd.read_csv(url_train)
# df_test = pd.read_csv(url_test)

# Guardar los datos para que queden disponibles siempre
# dir_train = 'C:/DS/_Portfolio/Python/Titanic_Clasificacion/titanic_train.csv'
# dir_test = 'C:/DS/_Portfolio/Python/Titanic_Clasificacion/titanic_test.csv'
# df_train.to_csv(dir_train)
# df_test.to_csv(dir_test)

# Importar fichero .csv con los datos
df_train = pd.read_csv('train.csv')
df_test = pd.read_csv('test.csv')

Visualicemos el contenido de ambos ficheros

In [7]:
print(df_train.head())
print(df_test.head())

   PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S  
  

 Podemos comprobar que ambos datasets contienen los mismos datos. El set de entrenamiento tiene una columna más, que indica si el pasajero sobrevivió o no. Con ésta información, podemos deducir que necesitaremos aplicar **algoritmos supervisados.** ya que para el entrenamiento dispongo de datos donde conozco el resultado.
 
 Además, nos damos cuenta de que faltan muchos datos en ambos set, por lo que será necesario realizar un pre-procesamiento de los datos.

## Entendiendo la información contenida en los datos
Muy útil para hacer el pre-procesamiento de los datos y los algoritmos que tendremos que aplicar.

- Verifico la cantidad de datos en ambos datasets
- Verifico el tipo de datos
- Verifico los datos faltantes
- Verifico las estadísticas de ambos datasets

### Cantidad de datos

In [4]:
df_train.shape

(891, 12)

In [8]:
df_test.shape

(418, 11)

Vemos que en el dataset de entrenamiento hay 891 registros con 12 columnas. En el de test, 418 registros y 11 columnas. Deducimos que el set de prueba tiene una columna menos ya que no indica si el pasajero sobrevivió o no. También deducimos que tenemos suficientes datos para entrenar nuestro modelo.

### Tipos de datos

In [6]:
df_train.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
PassengerId    891 non-null int64
Survived       891 non-null int64
Pclass         891 non-null int64
Name           891 non-null object
Sex            891 non-null object
Age            714 non-null float64
SibSp          891 non-null int64
Parch          891 non-null int64
Ticket         891 non-null object
Fare           891 non-null float64
Cabin          204 non-null object
Embarked       889 non-null object
dtypes: float64(2), int64(5), object(5)
memory usage: 66.2+ KB


In [7]:
df_test.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 418 entries, 0 to 417
Data columns (total 11 columns):
PassengerId    418 non-null int64
Pclass         418 non-null int64
Name           418 non-null object
Sex            418 non-null object
Age            332 non-null float64
SibSp          418 non-null int64
Parch          418 non-null int64
Ticket         418 non-null object
Fare           417 non-null float64
Cabin          91 non-null object
Embarked       418 non-null object
dtypes: float64(2), int64(4), object(5)
memory usage: 27.8+ KB


En ambos datasets tenemos tipos de datos numericos, flotantes y objetos. La columna Sex es de tipo objeto y lo tendremos que cambiar para poder utilizarla en los algoritmos.

### Comprobamos datos faltantes

In [9]:
pd.isnull(df_train).sum()

PassengerId      0
Survived         0
Pclass           0
Name             0
Sex              0
Age            177
SibSp            0
Parch            0
Ticket           0
Fare             0
Cabin          687
Embarked         2
dtype: int64

In [9]:
pd.isnull(df_test).sum()

PassengerId      0
Pclass           0
Name             0
Sex              0
Age             86
SibSp            0
Parch            0
Ticket           0
Fare             1
Cabin          327
Embarked         0
dtype: int64

In [19]:
filas_con_nulos = df_train.head().isnull().any(axis=1)
print(df_train.head()[filas_con_nulos])

   PassengerId  Survived  Pclass                      Name     Sex   Age  \
0            1         0       3   Braund, Mr. Owen Harris    male  22.0   
2            3         1       3    Heikkinen, Miss. Laina  female  26.0   
4            5         0       3  Allen, Mr. William Henry    male  35.0   

   SibSp  Parch            Ticket   Fare Cabin Embarked  
0      1      0         A/5 21171  7.250   NaN        S  
2      0      0  STON/O2. 3101282  7.925   NaN        S  
4      0      0            373450  8.050   NaN        S  


In [22]:
nan_rows = df_train[df_train['Cabin'].isna()]
print(nan_rows)


     PassengerId  Survived  Pclass                                      Name  \
0              1         0       3                   Braund, Mr. Owen Harris   
2              3         1       3                    Heikkinen, Miss. Laina   
4              5         0       3                  Allen, Mr. William Henry   
5              6         0       3                          Moran, Mr. James   
7              8         0       3            Palsson, Master. Gosta Leonard   
..           ...       ...     ...                                       ...   
884          885         0       3                    Sutehall, Mr. Henry Jr   
885          886         0       3      Rice, Mrs. William (Margaret Norton)   
886          887         0       2                     Montvila, Rev. Juozas   
888          889         0       3  Johnston, Miss. Catherine Helen "Carrie"   
890          891         0       3                       Dooley, Mr. Patrick   

        Sex   Age  SibSp  Parch        

In [25]:
# Seleccionar solo algunas columnas de las filas con 'Cabin' nulo
columnas_interes = ['Cabin', 'Name', 'Ticket']  # Cambia esto por las columnas que te interesan
nan_rows = df_train[df_train['Cabin'].isna()][columnas_interes]
print(nan_rows)

    Cabin                                      Name            Ticket
0     NaN                   Braund, Mr. Owen Harris         A/5 21171
2     NaN                    Heikkinen, Miss. Laina  STON/O2. 3101282
4     NaN                  Allen, Mr. William Henry            373450
5     NaN                          Moran, Mr. James            330877
7     NaN            Palsson, Master. Gosta Leonard            349909
..    ...                                       ...               ...
884   NaN                    Sutehall, Mr. Henry Jr   SOTON/OQ 392076
885   NaN      Rice, Mrs. William (Margaret Norton)            382652
886   NaN                     Montvila, Rev. Juozas            211536
888   NaN  Johnston, Miss. Catherine Helen "Carrie"        W./C. 6607
890   NaN                       Dooley, Mr. Patrick            370376

[687 rows x 3 columns]


Llama la atención la gran cantidad de datos faltantes en las columnas Age y Cabin. Cabin no afecta mucho al análisis. Age si es relevante. En el pre-rpocesamiento de datos, depuraremos ambos datasets.

### Estadísticas de cada dataset

In [10]:
df_train.describe()

PassengerId    Survived      Pclass         Age       SibSp  \
count   891.000000  891.000000  891.000000  714.000000  891.000000   
mean    446.000000    0.383838    2.308642   29.699118    0.523008   
std     257.353842    0.486592    0.836071   14.526497    1.102743   
min       1.000000    0.000000    1.000000    0.420000    0.000000   
25%     223.500000    0.000000    2.000000   20.125000    0.000000   
50%     446.000000    0.000000    3.000000   28.000000    0.000000   
75%     668.500000    1.000000    3.000000   38.000000    1.000000   
max     891.000000    1.000000    3.000000   80.000000    8.000000   

            Parch        Fare  
count  891.000000  891.000000  
mean     0.381594   32.204208  
std      0.806057   49.693429  
min      0.000000    0.000000  
25%      0.000000    7.910400  
50%      0.000000   14.454200  
75%      0.000000   31.000000  
max      6.000000  512.329200

In [11]:
df_test.describe()

PassengerId      Pclass         Age       SibSp       Parch        Fare
count   418.000000  418.000000  332.000000  418.000000  418.000000  417.000000
mean   1100.500000    2.265550   30.272590    0.447368    0.392344   35.627188
std     120.810458    0.841838   14.181209    0.896760    0.981429   55.907576
min     892.000000    1.000000    0.170000    0.000000    0.000000    0.000000
25%     996.250000    1.000000   21.000000    0.000000    0.000000    7.895800
50%    1100.500000    3.000000   27.000000    0.000000    0.000000   14.454200
75%    1204.750000    3.000000   39.000000    1.000000    0.000000   31.500000
max    1309.000000    3.000000   76.000000    8.000000    9.000000  512.329200

La media y la desviación estandard en ambos datasets son razonables. Podemos asumir que cualquier relación que encontremos en los datos de entrenamiento debería funcionar de manera similar en los de prueba.

## Preprocesamiento de los Datos
- Cambio los datos de sexos en numéros
- Cambio datos de embarque en números
- Reemplazo datos faltantes en la edad por el valor medio de la columna
- Creo varios grupos de acuerdo a franjas de edad
- Elimino la columna *cabin* ya que tiene muchos valores perdidos
- Elimino columnas innecesarias para el análisis
- Elimino filas con datos perdidos
- Verifico los datasets

#### Female -> 0, Male -> 1

In [12]:
df_train['Sex'].replace(['female', 'male'],[0,1], inplace = True)
df_test['Sex'].replace(['female', 'male'],[0,1], inplace = True)

#### Embarked Q -> 0, S -> 1, C -> 2

In [13]:
df_train['Embarked'].replace(['Q', 'S', 'C'], [0, 1, 2], inplace = True)
df_test['Embarked'].replace(['Q', 'S', 'C'], [0, 1, 2], inplace = True)

#### Reemplazo datos perdidos por el valor medio de la columna Age

In [14]:
df_train['Age'].mean()

29.69911764705882

In [15]:
df_test['Age'].mean()

30.272590361445783

Redondeo ambas medias a 30

In [16]:
promedio = 30
df_train['Age'] = df_train['Age'].replace(np.nan, promedio)
df_test['Age'] = df_test['Age'].replace(np.nan, promedio)

In [17]:
df_train.describe()

PassengerId    Survived      Pclass         Sex         Age  \
count   891.000000  891.000000  891.000000  891.000000  891.000000   
mean    446.000000    0.383838    2.308642    0.647587   29.758889   
std     257.353842    0.486592    0.836071    0.477990   13.002570   
min       1.000000    0.000000    1.000000    0.000000    0.420000   
25%     223.500000    0.000000    2.000000    0.000000   22.000000   
50%     446.000000    0.000000    3.000000    1.000000   30.000000   
75%     668.500000    1.000000    3.000000    1.000000   35.000000   
max     891.000000    1.000000    3.000000    1.000000   80.000000   

            SibSp       Parch        Fare    Embarked  
count  891.000000  891.000000  891.000000  889.000000  
mean     0.523008    0.381594   32.204208    1.102362  
std      1.102743    0.806057   49.693429    0.515181  
min      0.000000    0.000000    0.000000    0.000000  
25%      0.000000    0.000000    7.910400    1.000000  
50%      0.000000    0.000000   14.454200    1.000000  
75%      1.000000    0.000000   31.000000    1.000000  
max      8.000000    6.000000  512.329200    2.000000

#### Agrupo las edades en bandas (0-8, 9-15, 16-18, 19-25, 26-40, 41-60, 61-100)

In [18]:
bins = [0, 8, 15, 18, 25, 40, 60, 100] # rangos
names = ['1', '2', '3', '4', '5', '6', '7'] # nombres

df_train['Age'] = pd.cut(df_train['Age'], bins, labels = names)
df_test['Age'] = pd.cut(df_test['Age'], bins, labels = names)

#### Elimino la columna *Cabin* dado que tiene muchos datos perdidos

In [19]:
df_train.drop(['Cabin'], axis = 1, inplace = True)
df_test.drop(['Cabin'], axis = 1, inplace = True)

#### Elimino las columnas innecesarias para el análisis

In [20]:
df_train.drop(['PassengerId', 'Name', 'Ticket'], axis = 1, inplace = True)
df_test.drop(['Name', 'Ticket'], axis = 1, inplace = True) # No elimino PassengerId porque la utilizaré mas adelante

#### Eliminamos las filas con datos perdidos

In [21]:
df_train.dropna(axis = 0, how = 'any', inplace = True)
df_test.dropna(axis = 0, how = 'any', inplace = True)

#### Verifico los datasets

In [22]:
print(pd.isnull(df_train).sum())
print(pd.isnull(df_test).sum())

Survived    0
Pclass      0
Sex         0
Age         0
SibSp       0
Parch       0
Fare        0
Embarked    0
dtype: int64
PassengerId    0
Pclass         0
Sex            0
Age            0
SibSp          0
Parch          0
Fare           0
Embarked       0
dtype: int64


In [23]:
print(df_train.shape)
print(df_test.shape)

(889, 8)
(417, 8)


In [24]:
print(df_train.head())
print(df_test.head())

   Survived  Pclass  Sex Age  SibSp  Parch     Fare  Embarked
0         0       3    1   4      1      0   7.2500       1.0
1         1       1    0   5      1      0  71.2833       2.0
2         1       3    0   5      0      0   7.9250       1.0
3         1       1    0   5      1      0  53.1000       1.0
4         0       3    1   5      0      0   8.0500       1.0
   PassengerId  Pclass  Sex Age  SibSp  Parch     Fare  Embarked
0          892       3    1   5      0      0   7.8292         0
1          893       3    0   6      1      0   7.0000         1
2          894       2    1   7      0      0   9.6875         0
3          895       3    1   5      0      0   8.6625         1
4          896       3    0   4      1      1  12.2875         1


## Aplicación de Algoritmos de Machine Learning

- Separo columna con la información de los supervivientes
- Separo los datos de "train" en entrenamiento
- Prueba para probar algoritmos:
 - Regresión Logística
 - Máquinas de Vector Soporte
 - K Vecinos más cercanos
 
En ésta etapa utilizaremos los datos de entrenamiento.

In [25]:
# Separo la columna supervivientes
x = np.array(df_train.drop(['Survived'], 1)) # variables para construir el modelo
y = np.array(df_train['Survived']) # resultado

In [26]:
# Separo los datos del dataset "train"
x_train, x_test, y_train, y_test = train_test_split(x,y, test_size = 0.2)

In [27]:
# Regresion Logistica
logreg = LogisticRegression()
logreg.fit(x_train, y_train)
y_pred = logreg.predict(x_test)
print('Precisión Regresión Logística: ')
print(logreg.score(x_train, y_train))

Precisión Regresión Logística: 
0.810126582278481


In [28]:
# Máquina Vector Soporte
svc = SVC()
svc.fit(x_train, y_train)
y_pred = svc.predict(x_test)
print('Precisión Máquinas Vector Soporte: ')
print(svc.score(x_train, y_train))

Precisión Máquinas Vector Soporte: 
0.680731364275668


In [29]:
# Vecinos Mas Cercanos
knn = KNeighborsClassifier(n_neighbors = 3)
knn.fit(x_train, y_train)
y_pred = knn.predict(x_test)
print('Precisión Vecinos más Cercanos: ')
print(knn.score(x_train, y_train))

Precisión Vecinos más Cercanos: 
0.8635724331926864


Comprobamos que el algoritmo más adecuado para realizar las predicciones es **K Vecinos más cercanos**

## Prediccion utilizando los modelos y el dataset de prueba

In [30]:
# Creo una columna con los ids de los pasajeros y ver quién sobrevive.
ids = df_test['PassengerId']

In [31]:
# Prediccion Regresión Lineal
prediccion_logreg = logreg.predict(df_test.drop('PassengerId', axis = 1))
out_logreg = pd.DataFrame({'PassengerId':ids, 'Survived': prediccion_logreg})
print('Predicción Regresión Logística: ')
print(out_logreg.head())

Predicción Regresión Logística: 
   PassengerId  Survived
0          892         0
1          893         0
2          894         0
3          895         0
4          896         1


In [32]:
# Prediccion Maquinas Vector Soporte
prediccion_svc = svc.predict(df_test.drop('PassengerId', axis = 1))
out_svc = pd.DataFrame({'PassengerId':ids, 'Survived': prediccion_svc})
print('Predicción Máquins Vector Soporte: ')
print(out_svc.head())

Predicción Máquins Vector Soporte: 
   PassengerId  Survived
0          892         0
1          893         0
2          894         0
3          895         0
4          896         0


In [33]:
# Prediccion Vecinos mas Cercanos
prediccion_knn = knn.predict(df_test.drop('PassengerId', axis = 1))
out_knn = pd.DataFrame({'PassengerId':ids, 'Survived': prediccion_svc})
print('Predicción Vecinos Más Cercanos: ')
print(out_knn.head())

Predicción Vecinos Más Cercanos: 
   PassengerId  Survived
0          892         0
1          893         0
2          894         0
3          895         0
4          896         0
